cmd/snap-repair/runner.go

```diff
@@ -43,6 +46,91 @@ import (
 	"github.com/snapcore/snapd/strutil"
 )
 
+// Repair is a runnable repair.
+type Repair struct {
+	*asserts.Repair
+
+	run      *Runner
+	sequence int
+}
+
+// SetStatus sets the status of the repair in the state and saves the latter.
+func (r *Repair) SetStatus(status RepairStatus) {
+	brandID := r.BrandID()
+	cur := *r.run.state.Sequences[brandID][r.sequence-1]
+	cur.Status = status
+	r.run.setRepairState(brandID, cur)
+	r.run.SaveState()
+}
+
+// Run executes the repair script leaving execution trail files on disk.
+func (r *Repair) Run() error {
+	// write the script to disk
+	rundir := filepath.Join(dirs.SnapRepairRunDir, r.BrandID(), r.RepairID())
+	err := os.MkdirAll(rundir, 0775)
+	if err != nil {
+		return err
+	}
+
+	script := filepath.Join(rundir, fmt.Sprintf("script.r%d", r.Revision()))
+	err = osutil.AtomicWriteFile(script, r.Body(), 0700, 0)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now().Unix()
+	logPath := filepath.Join(rundir, fmt.Sprintf("r%d.%v.output", r.Revision(), now))
+	logf, err := os.Create(logPath)
+	if err != nil {
+		return err
+	}
+	defer logf.Close()
+
+	// run things and captures output etc
+	stR, stW, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+	defer stR.Close()
+	defer stW.Close()
+
+	// run the script
+	cmd := exec.Command(script)
+	cmd.ExtraFiles = []*os.File{stW}
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "SNAP_REPAIR_STATUS_FD=3")
+	cmd.Stdout = logf
+	cmd.Stderr = logf
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+	stW.Close()
+
+	// ignore error, we only care about what we got via the status pipe
+	cmd.Wait()
```

Comment:
should we have a timeout at some point?  at least a TODO about that here, maybe it needs to come specified from the repair itself?